x-pack/plugins/endpoint/server/routes/resolver.ts

```diff
@@ -0,0 +1,181 @@
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) => BaseSearchHandler;
+
+function createChildrenSearchHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new ChildrenSearchHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function createSingleNodeHandler(
+  client: IScopedClusterClient,
+  endpointContext: EndpointAppContext,
+  pageInfo: PaginationInfo,
+  countQuery: JSONish,
+  entityID: string
+) {
+  return new SingleNodeHandler(client, endpointContext, pageInfo, countQuery, entityID);
+}
+
+function getReqHandler(
+  searchQueryBuilder: SearchQueryBuilder,
+  countQueryBuilder: CountQueryBuilder,
+  debugQueryString: string,
+  notFoundString: string,
+  log: Logger,
+  endpointAppContext: EndpointAppContext,
+  handlerCreator: SearchHandlerCreator
+): RequestHandler<unknown, ResolverRequestParams> {
+  return async (context, req, res) => {
+    const entityID = req.query.entity_id;
+    log.debug(`entity_id: ${entityID}`);
+    const paginationInfo = {
+      page: req.query?.page_index,
+      pageSize: req.query?.page_size,
+    } as PaginationInfo;
+    try {
+      const query = await searchQueryBuilder(endpointAppContext, entityID, paginationInfo);
+      log.debug(`${debugQueryString} ${JSON.stringify(query)}`);
+      const response = (await context.core.elasticsearch.dataClient.callAsCurrentUser(
+        'search',
+        query
+      )) as SearchResponse<ResolverData>;
+
+      if (response.hits.hits.length === 0) {
+        return res.notFound({ body: `${notFoundString}` });
+      }
+      const countQuery = countQueryBuilder(entityID).query;
+      const handler = handlerCreator(
+        context.core.elasticsearch.dataClient,
+        endpointAppContext,
+        paginationInfo,
+        countQuery,
+        entityID
+      );
+      return res.ok({
+        body: await handler.buildResponse(
+          response.hits.hits,
+          // total is an object in kibana >=7.0 so cast it to an object
+          // see https://github.com/elastic/kibana/issues/56694
+          (response.hits.total as unknown) as Total
+        ),
+      });
+    } catch (err) {
+      return handleError(res, log, err);
+    }
+  };
+}
+
+export function registerResolverRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  const log = endpointAppContext.logFactory.get('resolver');
+  const validateQueryObject = schema.object({
+    page_size: schema.maybe(schema.number({ min: 1, max: 1000 })),
+    page_index: schema.maybe(schema.number({ min: 0 })),
+    entity_id: schema.string(),
+  });
+
+  router.get(
+    {
+      path: '/api/endpoint/resolver/children',
+      validate: {
+        query: validateQueryObject,
+      },
+      options: { authRequired: true },
+    },
+    getReqHandler(
```

Comment:
@jonathan-buttner I have an idea that may make this code a little bit easier to follow. Please bear with me. Since there is a class called `BaseSearchHandler` and two classes that extend it `ChildrenSearchHandler` and `SingleNodeHandler`, I was thinking that `BaseSearchHandler` could have an abstract method called `handleRequest()`. If we did that this code would become:
```javascript
router.get({...}, ChildrenSearchHandler.handleRequest(endpointAppContext))
```
Internally, the `ChildrenSearchHandler` could define its own `searchQueryBuilder()`, `countQueryBuilder()` and `buildResponse()` methods. ChildrenSearchHandler.handleRequest` would use these methods and look something like:
```javascript
  export async function handleRequest(...) {
      // Uses internal searchQueryBuilder() function
      const query = await this.searchQueryBuilder();
      const response = (await context.core.elasticsearch.dataClient.callAsCurrentUser(
        'search',
        query
      )) as SearchResponse<ResolverData>;

      // Uses internal countQueryBuilder() function
      const countQuery = this.countQueryBuilder(entityID).query;

      // Uses internal buildResponse() function
      return res.ok({
        body: await this.buildResponse(
          response.hits.hits,
          (response.hits.total as unknown) as Total
        ),
      });
  }
```
I feel like this way, we move closer to using the inheritance model. Right now we have a mix of composition and inheritance. We could also move closer to the composition model instead. I just think that having a mix makes it harder to follow. @oatkiller @marshallmain Any thoughts?

Reply:
That's a great idea Pedro, thanks for pointing that out. I'll probably move the `handleRequest` method into `BaseSearchHandler` since it'll be pretty similar in both child classes.

Reply:
@peluja1012 @jonathan-buttner Since we only have 2 classes inheriting from BaseSearchHandler I would favor keeping handleRequest abstract in BaseSearchHandler and implementing it in each child class even if they are identical.  I think the logic that belongs in the base class is the pagination logic.  Overall I think moving closer to the inheritance model would help but there's more we can do to simplify the API endpoints here, I left some other comments with my thoughts.